pyccel/ast/builtins.py

```diff
@@ -774,6 +775,40 @@ def is_homogeneous(self):
         """
         return True
 
+
+class PythonSetFunction(PyccelFunction):
+    """
+    Class representing a call to the `set` function.
+
+    Class representing a call to the `set` function. This is
+    different to the `{,}` syntax as it only takes one argument
+    and unpacks any variables.
+
+    Parameters
+    ----------
+    arg : TypedAstNode
+        The argument passed to the function call.
+    """
+
+    __slots__ = ('_shape', '_class_type')
+    name = 'set'
+    def __new__(cls, arg):
+        if isinstance(arg.class_type, HomogeneousSetType):
+            return arg
+        elif isinstance(arg, PythonList):
+            return PythonSet(*arg)
+        elif isinstance(arg.shape[0], LiteralInteger):
+            return PythonSet(*[arg[i] for i in range(arg.shape[0])])
```

Comment:
A `PythonList` has a shape which is a `LiteralInteger`. Why do you feel that it is important to treat the case of a `PythonList` individually but not a `PythonTuple` or a `PythonSet`?

Reply:
Do you think it should be like this: 
```suggestion
        elif isinstance(arg, (PythonList, PythonSet, PythonTuple)):
            return PythonSet(*arg))
```

Reply:
Actually I was thinking of:
```suggestion
        elif isinstance(arg.shape[0], LiteralInteger):
            return PythonSet(*[arg[i] for i in range(arg.shape[0])])
```
which should be the most general and keep working for future types, but it also depends on what translation you want for `Variable`s of known length

Reply:
I would be fine with either solution

Reply:
I got an error when i tried this :
```python 
k = set([1,2,3,4])
```
This is the error: `|fatal [semantic]: test.py [2,4]| 'PythonList' object is not subscriptable (set((1, 2, 3, 4)))`

Reply:
Fixed : 4576d39